### PR TITLE
[DOC] initial operators commit

### PR DESCRIPTION
**Pull request by shmolf**
This Pull Request seeks to include documentation surrounding operators and how they behave on a high level.

### CODE REVIEW

**Comment by shmolf** [comment]
I did not mean to yet open a PR in the main repo. My apologies.

**Comment by nobu** [comment]
You could turn this into "draft".

File: doc/syntax/operators.rdoc [hunk 1]
```diff
@@ -0,0 +1,22 @@
+= Operators
+
+In Ruby, operators such as +++, are defined as methods on the class.
```
Comment (nobu) [inline doc/syntax/operators.rdoc:3]: `+++` doesn't seem to work.

```suggestion
In Ruby, operators such as <code>+</code>, are defined as methods on the class.
```

File: doc/syntax/operators.rdoc [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+Literals[link:/literals.rdoc] define their methods within the lower level, C language.
```
Comment (nobu) [inline doc/syntax/operators.rdoc:4]: RDoc embeds `link:` as-is.

```suggestion
Literals[rdoc-ref:syntax/literals.rdoc] define their methods within the lower level, C language.
```

File: doc/syntax/operators.rdoc [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+{String Class}[rdoc-ref:string.c], for example.
```
Comment (nobu) [inline doc/syntax/operators.rdoc:5]: Source code files will not be included generated files.
Bare class name is linked to the page for that class, instead.

```suggestion
String class, for example.
```

File: doc/syntax/operators.rdoc [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+Non-literal objects can define or overload their own implementation for a given operator.
```
Comment (nobu) [inline doc/syntax/operators.rdoc:7]: This sentence feels ambiguous.
What do you mean by "non-literal objects" here?
Non-builtin class as the following?
Or, regarding that singleton methods cannot be defined on literals?

File: doc/syntax/operators.rdoc (hunk 1 above)
Comment (shmolf) [reply doc/syntax/operators.rdoc]: Thank you for that suggestion.

File: doc/syntax/operators.rdoc (hunk 4 above)
Comment (shmolf) [reply doc/syntax/operators.rdoc]: I was thinking of Ruby objects. You're right, `Non-literal` is very ambiguous. pushed up a fix.

File: doc/syntax/operators.rdoc (hunk 3 above)
Comment (shmolf) [reply doc/syntax/operators.rdoc]: Good to know! I'll remove the link refernce.

File: doc/syntax/operators.rdoc (hunk 2 above)
Comment (shmolf) [reply doc/syntax/operators.rdoc]: Thank you for this suggestion.

File: doc/syntax/operators.rdoc [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+In Ruby, operators such as <code>+</code>, are defined as methods on the class.
+Literals[rdoc-ref:syntax/literals.rdoc] define their methods within the lower
+level, C language. String class, for example.
+
+Ruby objects can define or overload their own implementation for most operators.
+
+Here is an example:
+
+class Foo < String
+    def &&(exp)
```
Comment (nobu) [inline doc/syntax/operators.rdoc:12]: `&&` operator can not be a method.

Comment (shmolf) [reply doc/syntax/operators.rdoc:12]: !!!

I forgot to remove this when I discovered it was tokenized.

Thank you

File: doc/syntax/operators.rdoc [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+== Operator Behavior
+
+How a class behaves to a given operator is specific to that class, since
+operators are method implementations.
+
+When using an operator, it's the expression on the left-hand side of the
+operation that specifies the behavior.
+
+  'a' * 3         # aaa
```
Comment (nobu) [inline doc/syntax/operators.rdoc:34]: Usually, we are using `#=>` for normal return values, and should quote to show it is a string.

```suggestion
  'a' * 3         #=> "aaa"
```

File: doc/syntax/operators.rdoc [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+Some examples:
+
+  true && 9 && "string"                       ## "string"
+  (1 + 2) && nil && "string"                  ## nil
+  (a = 1) && (b = false) && (c = "string")    ## false
+
+  puts a                                      ## 1
+  puts b                                      ## false
+  puts c                                      ## nil
```
Comment (nobu) [inline doc/syntax/operators.rdoc:62]:
```suggestion
  true && 9 && "string"                       #=> "string"
  (1 + 2) && nil && "string"                  #=> nil
  (a = 1) && (b = false) && (c = "string")    #=> false

  puts a                                      #=> 1
  puts b                                      #=> false
  puts c                                      #=> nil
```

File: doc/syntax/operators.rdoc [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+=== <code>||</code>
+
+The means by which <code>||</code> short-circuits, is to return the result of
+the first expression that is truthy.
+
+Some examples:
+
+  (1 + 2) || true || "string"                 ## 3
+  false || nil || "string"                    ## "string"
```
Comment (nobu) [inline doc/syntax/operators.rdoc:74]:
```suggestion
  (1 + 2) || true || "string"                 #=> 3
  false || nil || "string"                    #=> "string"
```

File: doc/syntax/operators.rdoc [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+  3 * 'a'         # TypeError: String can't be coerced into Integer
+
+== Logical Operators
+
+Logical operators are not methods on the Ruby Object, and therefor cannot be
+redefined/overloaded. They are tokenized at a lower level.
+
+Short-circuit logical operators, <code>&&</code> and <code>||</code>, do not
+always result in a boolean value. Similar to blocks, it's the last evaluated
+expression that defines the result of the operation.
```
Comment (nobu) [inline doc/syntax/operators.rdoc:44]: Same for `and` and `or`.
Isn't it needed?

Comment (shmolf) [reply doc/syntax/operators.rdoc]: I think it makes sense to include those.

Commit pushed

**Comment by shmolf** [comment]
@nobu Any preference for me to squash these commits before approval, or as part of the Merge?

File: doc/syntax/operators.rdoc [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+When using an operator, it's the expression on the left-hand side of the
+operation that specifies the behavior.
+
+  'a' * 3         #=> "aaa"
+  3 * 'a'         # TypeError: String can't be coerced into Integer
+
+== Logical Operators
+
+Logical operators are not methods on the Ruby Object, and therefor cannot be
```
Comment (nobu) [inline doc/syntax/operators.rdoc:39]: Do you intend this `Object` as `::Object` class?

Comment (jeremyevans) [reply doc/syntax/operators.rdoc:39]: This would also be true of `BasicObject`.  I recommend removing the ` on the Ruby Object` part.

Comment (shmolf) [reply doc/syntax/operators.rdoc]: Pushed up a commit.

**Comment by nobu** [comment]
> @nobu Any preference for me to squash these commits before approval, or as part of the Merge?

Either is fine.

**Comment by shmolf** [comment]
@nobu I wanted to double-check if this Pull Request is waiting on any action from me.

**Pull request merged**